src/layers/SoftMax.jl

```diff
@@ -1,41 +1,56 @@
 type SoftMax <: Nonlinearity
-    last_input  :: Array{Float64}
-    last_output :: Array{Float64}
-    last_loss   :: Array{Float64}
+    x  :: Array{Float64}        # input m by n matrix which represents m instances with scores for n classes
+    y :: Array{Float64}         # output m by n matrix which uses softmax to normalize the input matrix
+    has_init :: Bool            # true if the layer has been initialized
+    jacobian :: Array{Float64}  # cache for the jacobain matrices used in backward
+    dldx :: Array{Float64}      # cahce for the backward result
 
     function SoftMax()
-        return new(Float64[], Float64[], Float64[])
+        return new(Float64[], Float64[], false, Float64[], Float64[])
     end
 end
 
-function forward(l::SoftMax, X::Array{Float64}; kwargs...)
-	X = broadcast(+, X, -maximum(X))
-    l.last_input = X
-    sumX = sum(exp(X))
-    Y = (x ./ sumX)
-    l.last_output = Y
-    return Y
+function init(l::SoftMax, p::Union{Layer,Void}, config::Dict{String,Any}; kwargs...)
+  if p == nothing
+      @assert ndims(config["input_size"]) == 1 # TODO: maybe an error message?
+      out_size = (config["batch_size"], config["input_size"][1])
+  else
+      out_size = getOutputSize(p)
+  end
+  l.x = Array{Float64}(out_size)
+  l.y = Array{Float64}(out_size)
+  l.has_init = true;
+  m,n = out_size
+  l.jacobian = Array{Float64}(n,n)
+  l.dldx = Array{Float64}(out_size)
+end
+
+function forward(l::SoftMax, X::Array{Float64,2}; kwargs...)
+
+    l.x = X
+    # iterating each row/picture
+    lxexp = exp(l.x)
```

Comment:
Both `Softmax` and `CrossEntropy` are now reimplemented in the most concise fashion with preallocations also been taken care. Dot operations though, i.e. dot division in `SoftMax.jl` line 39 and dot multiplication `CrossEntropyLoss.jl` line 36, takes unreasonable allocations even though I preallocate their output, especially for the latter, where the negation `-` takes same amount of allocations as the `.*`. Looking for a more efficient way here.